### PR TITLE
[CPU] Integrate i8mm patterns from upstream

### PR DESCRIPTION
**Pull request by KoolJBlack**
Inserts `populateLowerVectorToArmNeonPatterns()` to trigger smmla arm neon instructions on compatible tile sizes.

benchmark-extra: android-cpu-dt-only

### CODE REVIEW

**Comment by KoolJBlack** [comment]
There is a fix upstream for the breakage here in https://github.com/llvm/llvm-project/pull/88288. Once this is cherry picked in, test should pass.

**Review by dcaballe — COMMENTED** [review]
LG but we have to wait for https://github.com/llvm/llvm-project/pull/88591/ to be integrated, just to make sure we don't generate wrong code.

This PR should also set the "use mmt4d custom kernels" flag that you added to false by default.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVirtualVectorLowering.cpp [hunk 1]
```diff
@@ -53,6 +56,7 @@ void LLVMCPUVirtualVectorLoweringPass::runOnOperation() {
           .setVectorTransferSplit(vectorTransferSplit);
 
   RewritePatternSet patterns(ctx);
+  arm_neon::populateLowerContractionToSMMLAPatternPatterns(patterns);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVirtualVectorLowering.cpp]: I think this should only be applied when we target arm_neon. We can provide options to the pass. The start point should be below: we look at target attributes, and determine if we enable a set of patterns. Then we add options to LLVMCPUVirtualVectorLowering, and set the option based on LLVMCPUPipelineOptions. LLVMCPUVectorTransposeLowering pass is an example.

https://github.com/openxla/iree/blob/503edab02257a42e865d08fb3d7d06263c0b62a3/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp#L96-L106

Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVirtualVectorLowering.cpp]: [Edited]

I see your point. Let me try adding options to the `LLVMCPUVirtualVectorLoweringPass` so this only activates on arm neon. I've also updated the tile sizes in CPUMaterializeEncodings to also look for the neon target option. 

@dcaballe i8mm can only be used with neon? I couldn't find a definitive answer in the documentation, but in my case I've only tested it on neon. Outside of CPUMaterializeEncodings, there are no later checks on neon specifically before adding the patterns.

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVirtualVectorLowering.cpp]: `+neon` doesn't include i8mm. You may want to add a `hasI8mmFeature` utility that checks for `+i8mm`. I'm not sure if there is a specific i8mm feature for SVE. @banach-space should know but we may want to enable it for SVE as well if that's not the case.

Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVirtualVectorLowering.cpp]: Updated `LLVMCPUVirtualVectorLoweringPass` to use the pipeline options. these patterns are now gated on the `+neon` cpu features.

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVirtualVectorLowering.cpp]: Sorry if my previous comment wasn't clear: we should check for `+i8mm`, not `+neon`.

Comment (banach-space) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVirtualVectorLowering.cpp]: > I'm not sure if there is a specific i8mm feature for SVE.

It's the same name as for NEON :) Btw, nice to see this happening @KoolJBlack  🙏🏻

Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUVirtualVectorLowering.cpp]: switch to i8mm

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp [hunk 2]
```diff
@@ -104,6 +104,9 @@ void LLVMCPULowerExecutableTargetPass::runOnOperation() {
   pipelineOpts.enableUkernels = hasUkernel(target);
   pipelineOpts.enableAArch64SSVE =
       isAArch64(target) && hasAnySVEFeature(target) && hasSMEFeature(target);
+  pipelineOpts.enableAArch64Neon = isAArch64(target) && hasNeonFeature(target);
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp]: I think we shouldn't need like 107

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp [hunk 3]
```diff
@@ -104,6 +104,9 @@ void LLVMCPULowerExecutableTargetPass::runOnOperation() {
...
+  pipelineOpts.enableAArch64I8mm =
+      isAArch64(target) && hasI8mmFeature(target) && hasNeonFeature(target);
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp]: I would also remove `hasNeonFeature` from 109. `+i8mm` includes `+neon` but we are not expanding all the features in IREE.

**Review by dcaballe — APPROVED** [review]
One small fix and should be good to go!

File: compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp:1]: The PR is not only enable i8mm patterns in vector lowering, but also updates the logics about data-tiling. Can you update the PR title and description?

File: compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp [hunk 4]
```diff
@@ -159,7 +159,17 @@ enumerateMatmulTileArm64(TypeRange elementTypes, ExecutableTargetAttr target) {
 
   if (!hasUkernel(target)) {
     if (lhs.isSignlessInteger(8) && rhs.isSignlessInteger(8) &&
-        (out.isSignlessInteger(32) || out.isF32())) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp:164]: what happens to f32?

File: compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp [hunk 5]
```diff
@@ -168,7 +178,16 @@ enumerateMatmulTileArm64(TypeRange elementTypes, ExecutableTargetAttr target) {
       };
     }
     if (lhs.isSignlessInteger(8) && rhs.isSignlessInteger(4) &&
-        (out.isSignlessInteger(32) || out.isF32())) {
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp:173]: what happens to f32?

**Review by hanhanW — CHANGES_REQUESTED** [review]
This seems to disable data-tiling for i8.i8.f32. Is it expected?

File: compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp (hunk 4 above)
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp:164]: This is probably my fault as I copied these tile sizes from the ukernel version. We can restore it and revisit separately to make sure any changes to the f32 cases don't add any noise. We shouldn't be using i8mm tile sizes for f32 as i8mm is not used for those cases. We want the f32 cases to fall into the default case.

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp:164]: I see, @KoolJBlack can you add it to the PR description which explains why f32 is removed?

**Review by hanhanW — APPROVED** [review]

File: compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp (hunk 4 above)
Comment (KoolJBlack) [reply compiler/src/iree/compiler/Codegen/Common/CPU/CPUMaterializeEncodingPass.cpp:164]: done

**Comment by dcaballe** [comment]
Hey @KoolJBlack. I fixed a few things but it looks like there's still something that is not triggering for some cases. See these benchmark results, for example:
```
matmul_1x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 0.079 (1.0X) | 0.069 (1.2X) | 0.016 (5.1X)
-- | -- | -- | --
matmul_1x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 0.072 (1.0X) | 0.065 (1.1X) | 0.016 (4.4X)
matmul_256x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 12.138 (1.0X) | 10.073 (1.2X) | 1.409 (8.6X)
matmul_256x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 16.543 (1.0X) | 6.788 (2.4X) | 1.187 (13.9X)

```matmul_1x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[0.079 (1.0X)](https://perf.iree.dev/serie?IREE?0633a7b6344c075e248724b36277bf8573bc8ec908035edd3d3e0bbf5e54d665)	[0.069 (1.2X)](https://perf.iree.dev/serie?IREE?e4b89492467994aef5aa22a4703d66f31439e402cbf5c0c97b83c5dd31f8f961)	[0.016 (5.1X)](https://perf.iree.dev/serie?IREE?f1aadd5288ea52756315a12789fca4230df1acf750fcf31445c87fcfb23be1d7)
matmul_1x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[0.072 (1.0X)](https://perf.iree.dev/serie?IREE?605e091d941596ebf44e0238fd1f791285ac7e2f33fff0110991c2a329166ade)	[0.065 (1.1X)](https://perf.iree.dev/serie?IREE?b2c090681bafdea5b2b2c9e22e9aff72341e918bff836535ab0f5ba74281c293)	[0.016 (4.4X)](https://perf.iree.dev/serie?IREE?cfa03d0422069dfa5f9483ab518a29261140d9bfb42035dd17018a8b5f1e89ca)
matmul_256x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[12.138 (1.0X)](https://perf.iree.dev/serie?IREE?a694805fd2aa24f7bb3464e817ade1eda09588928e5b168947eef7e6b5ac8dee)	[10.073 (1.2X)](https://perf.iree.dev/serie?IREE?e8065d1ddf99eb564720a2efd4c3b45e2553b6b34f71a857501a49c39c1d8f82)	[1.409 (8.6X)](https://perf.iree.dev/serie?IREE?fe0a953188f398da446a84e74ad069d4029568c0a02709b84bef8922533bb14a)
matmul_256x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[16.543 (1.0X)](https://perf.iree.dev/serie?IREE?6619268e9e04aeba690b0385f6ae2374127bee5ec06a0b06c3c1229da8b48428)	[6.788 (2.4X)](https://perf.iree.dev/serie?IREE?11a6e99b693ab79387b99f83a006c5d100d03842fe5c69ce4afd8acd7eb71803)	[1.187 (13.9X)](https://perf.iree.dev/serie?IREE?3230d9aca0ea36e75ea0ad6d604939a5ae70b0c6dc5a8049b20fbe98d509aefa)
```

Would you mind taking a look?

**Comment by KoolJBlack** [comment]
> Hey @KoolJBlack. I fixed a few things but it looks like there's still something that is not triggering for some cases. See these benchmark results, for example:
> 
> ```
> matmul_1x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 0.079 (1.0X) | 0.069 (1.2X) | 0.016 (5.1X)
> -- | -- | -- | --
> matmul_1x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 0.072 (1.0X) | 0.065 (1.1X) | 0.016 (4.4X)
> matmul_256x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 12.138 (1.0X) | 10.073 (1.2X) | 1.409 (8.6X)
> matmul_256x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 16.543 (1.0X) | 6.788 (2.4X) | 1.187 (13.9X)
> 
> ```matmul_1x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[0.079 (1.0X)](https://perf.iree.dev/serie?IREE?0633a7b6344c075e248724b36277bf8573bc8ec908035edd3d3e0bbf5e54d665)	[0.069 (1.2X)](https://perf.iree.dev/serie?IREE?e4b89492467994aef5aa22a4703d66f31439e402cbf5c0c97b83c5dd31f8f961)	[0.016 (5.1X)](https://perf.iree.dev/serie?IREE?f1aadd5288ea52756315a12789fca4230df1acf750fcf31445c87fcfb23be1d7)
> matmul_1x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[0.072 (1.0X)](https://perf.iree.dev/serie?IREE?605e091d941596ebf44e0238fd1f791285ac7e2f33fff0110991c2a329166ade)	[0.065 (1.1X)](https://perf.iree.dev/serie?IREE?b2c090681bafdea5b2b2c9e22e9aff72341e918bff836535ab0f5ba74281c293)	[0.016 (4.4X)](https://perf.iree.dev/serie?IREE?cfa03d0422069dfa5f9483ab518a29261140d9bfb42035dd17018a8b5f1e89ca)
> matmul_256x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[12.138 (1.0X)](https://perf.iree.dev/serie?IREE?a694805fd2aa24f7bb3464e817ade1eda09588928e5b168947eef7e6b5ac8dee)	[10.073 (1.2X)](https://perf.iree.dev/serie?IREE?e8065d1ddf99eb564720a2efd4c3b45e2553b6b34f71a857501a49c39c1d8f82)	[1.409 (8.6X)](https://perf.iree.dev/serie?IREE?fe0a953188f398da446a84e74ad069d4029568c0a02709b84bef8922533bb14a)
> matmul_256x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[16.543 (1.0X)](https://perf.iree.dev/serie?IREE?6619268e9e04aeba690b0385f6ae2374127bee5ec06a0b06c3c1229da8b48428)	[6.788 (2.4X)](https://perf.iree.dev/serie?IREE?11a6e99b693ab79387b99f83a006c5d100d03842fe5c69ce4afd8acd7eb71803)	[1.187 (13.9X)](https://perf.iree.dev/serie?IREE?3230d9aca0ea36e75ea0ad6d604939a5ae70b0c6dc5a8049b20fbe98d509aefa)
> ```
> 
> Would you mind taking a look?

I check the models in the benchmark by running local. I can confirm the patterns are intact properly activating on all cases. These ones however, aren't showing speedups for some reason. I'm profiling locally to figure out where the differences is, especially since there are clearly improvements in other benchmarks and I only got improved numbers from my own test cases. In the meantime, I can't find anything specifically wrong (I've been checking) so i'm committing this.

**Comment by KoolJBlack** [comment]
> > Hey @KoolJBlack. I fixed a few things but it looks like there's still something that is not triggering for some cases. See these benchmark results, for example:
> > ```
> > matmul_1x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 0.079 (1.0X) | 0.069 (1.2X) | 0.016 (5.1X)
> > -- | -- | -- | --
> > matmul_1x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 0.072 (1.0X) | 0.065 (1.1X) | 0.016 (4.4X)
> > matmul_256x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 12.138 (1.0X) | 10.073 (1.2X) | 1.409 (8.6X)
> > matmul_256x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores] | 16.543 (1.0X) | 6.788 (2.4X) | 1.187 (13.9X)
> > 
> > ```matmul_1x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[0.079 (1.0X)](https://perf.iree.dev/serie?IREE?0633a7b6344c075e248724b36277bf8573bc8ec908035edd3d3e0bbf5e54d665)	[0.069 (1.2X)](https://perf.iree.dev/serie?IREE?e4b89492467994aef5aa22a4703d66f31439e402cbf5c0c97b83c5dd31f8f961)	[0.016 (5.1X)](https://perf.iree.dev/serie?IREE?f1aadd5288ea52756315a12789fca4230df1acf750fcf31445c87fcfb23be1d7)
> > matmul_1x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[0.072 (1.0X)](https://perf.iree.dev/serie?IREE?605e091d941596ebf44e0238fd1f791285ac7e2f33fff0110991c2a329166ade)	[0.065 (1.1X)](https://perf.iree.dev/serie?IREE?b2c090681bafdea5b2b2c9e22e9aff72341e918bff836535ab0f5ba74281c293)	[0.016 (4.4X)](https://perf.iree.dev/serie?IREE?cfa03d0422069dfa5f9483ab518a29261140d9bfb42035dd17018a8b5f1e89ca)
> > matmul_256x256x2048_i8_i4_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[12.138 (1.0X)](https://perf.iree.dev/serie?IREE?a694805fd2aa24f7bb3464e817ade1eda09588928e5b168947eef7e6b5ac8dee)	[10.073 (1.2X)](https://perf.iree.dev/serie?IREE?e8065d1ddf99eb564720a2efd4c3b45e2553b6b34f71a857501a49c39c1d8f82)	[1.409 (8.6X)](https://perf.iree.dev/serie?IREE?fe0a953188f398da446a84e74ad069d4029568c0a02709b84bef8922533bb14a)
> > matmul_256x256x2048_i8_i8_i32_tile_config_default(linalg) [armv8.2-a-generic-linux_android29-llvm_cpu] local_sync(embedded_elf)[full-inference,default-flags] with default @ pixel-6-pro[big-cores]	[16.543 (1.0X)](https://perf.iree.dev/serie?IREE?6619268e9e04aeba690b0385f6ae2374127bee5ec06a0b06c3c1229da8b48428)	[6.788 (2.4X)](https://perf.iree.dev/serie?IREE?11a6e99b693ab79387b99f83a006c5d100d03842fe5c69ce4afd8acd7eb71803)	[1.187 (13.9X)](https://perf.iree.dev/serie?IREE?3230d9aca0ea36e75ea0ad6d604939a5ae70b0c6dc5a8049b20fbe98d509aefa)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Would you mind taking a look?
> 
> I check the models in the benchmark by running local. I can confirm the patterns are intact properly activating on all cases. These ones however, aren't showing speedups for some reason. I'm profiling locally to figure out where the differences is, especially since there are clearly improvements in other benchmarks and I only got improved numbers from my own test cases. In the meantime, I can't find anything specifically wrong (I've been checking) so i'm committing this.

So it turns out I was looking for the issue in the wrong place. The reason for the slower then expected performance on the CLI is due to matmul microbenchmarks not using the `+i8mm` & `+neon` cpu features, which are required to activate the fast path. Current config is only `--iree-llvmcpu-target-cpu-features=+dotprod`. 

I've actually gotten 1-2ms times on Pixel 8 Pro inline with the DT+UK path. Original assumption was difference in perf was due to faster phone, but digging through CLI shows why my results differ. 

@ScottTodd Is there a reason for the for the current CLI flags or any reason I cannot add more in a separate PR for faster benchmarks?

**Comment by ScottTodd** [comment]
Do you mean "CI" (continuous integration)? (not sure what you mean by "CLI)

> @ScottTodd Is there a reason for the for the current CLI flags or any reason I cannot add more in a separate PR for faster benchmarks?

For which? Can you link to some code? I'm not very familiar with the matmul benchmarks you seem to be referencing. We generally aim to test "default flags" and then some bucket(s) of experimental/optional/performance flags. Host platforms can use `--iree-llvmcpu-target-cpu-features=host`, but targets like Android currently need to specify the features explicitly.

**Pull request merged**